scipy/stats/_qmc.py

```diff
@@ -481,6 +485,186 @@ def _perturb_discrepancy(sample: np.ndarray, i1: int, i2: int, k: int,
     return disc_ep
 
 
+def _jitter_sample(
+        sample: np.ndarray,
+        rng: Union[np.random.Generator, np.random.RandomState],
+        eps: float = 1e-10
+) -> np.ndarray:
+    """Randomly jitter `sample` until they are unique.
+
+    If the samples are not unique, the number of regions in our
+    tessellation will be less than the number of input samples.
+    """
+    vals, count = np.unique(sample, return_counts=True)
+    if np.any(vals[count > 1]):
+        offset = rng.uniform(-eps, eps, size=(len(sample), sample.shape[1]))
+        sample = sample * (1 + offset)
+    return sample
+
+
+def _lloyd_centroidal_voronoi_tessellation(
+        sample: np.ndarray,
+        decay: float,
+        rng: Union[np.random.Generator, np.random.RandomState]
+) -> np.ndarray:
+    """Lloyd's algorithm iteration.
+
+    Based on the implementation of Stéfan van der Walt:
+
+    https://github.com/stefanv/lloyd
+
+    which is:
+
+        Copyright (c) 2021-04-21 Stéfan van der Walt
+        https://github.com/stefanv/lloyd
+        MIT License
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on.
+    decay : float
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere.
+    rng : `numpy.random.Generator`
+        Random number generator.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        The sample after an iteration of Lloyd's algorithm.
+
+    """
+    sample = _jitter_sample(sample, rng=rng)
+    new_sample = np.empty_like(sample)
+
+    voronoi = spatial.Voronoi(sample)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move sample towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_sample[ii] = sample[ii] + (centroid - sample[ii]) * decay
+
+    # only update sample to centroid within the region
+    is_valid = np.all(np.logical_and(new_sample >= 0, new_sample <= 1), axis=1)
+    sample[is_valid] = new_sample[is_valid]
+
+    return sample
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        sample: npt.ArrayLike,
+        *,
+        n_iters: IntNumber = 10,
```

Comment:
for `n_iters`, do you want to add error handling for values `<0` and non-integer:

The first case below returns the input `sample` unchanged (that *might* be "ok," though slighently allowing negative iterations is perhaps a bit weird?).

The second case produces a long traceback that includes:

```
        self       = <scipy.stats.tests.test_qmc.TestUtils object at 0x10d8dd2e0>
scipy/stats/_qmc.py:657: in lloyd_centroidal_voronoi_tessellation
    decay = (np.exp(-x / root) + 0.9 for x in range(n_iters))
E   TypeError: 'float' object cannot be interpreted as an integer
```

That's easy enough to catch early I suspect

Reply:
for this draft test:

```diff
diff --git a/scipy/stats/tests/test_qmc.py b/scipy/stats/tests/test_qmc.py
index a5346d885..7ee6804e4 100644
--- a/scipy/stats/tests/test_qmc.py
+++ b/scipy/stats/tests/test_qmc.py
@@ -362,6 +362,15 @@ class TestUtils:
             qmc.lloyd_centroidal_voronoi_tessellation(sample, seed=SEED)
 
 
+    def test_lloyd_varied_input(self):
+        rng = np.random.RandomState(1809831)
+        sample = rng.uniform(0, 1, size=(80, 3))
+        new_sample = qmc.lloyd_centroidal_voronoi_tessellation(sample, n_iters=-3, seed=SEED)
+        assert_allclose(new_sample, sample)
+
+        qmc.lloyd_centroidal_voronoi_tessellation(sample, n_iters=2.17, seed=SEED)
+
+
```

Reply:
`n_iters` is no longer a parameter, but along these same lines, it looks like we could use some more input validation tests. For instance, what happens now when `maxiter` is negative or a float?